examples/pmatch.kl

```diff
@@ -0,0 +1,145 @@
+#lang "n-ary-app.kl"
+(import "defun.kl")
+(import "pair-datatype.kl")
+(import "list-datatype.kl")
+
+(import (shift "n-ary-app.kl" 1))
+
+(import (shift "quasiquote.kl" 1))
+(import (shift "defun.kl" 1))
+(import (shift "lispy-do.kl" 1))
+(import (shift "free-identifier-case.kl" 1))
+(import (shift "temporaries.kl" 1))
+(import (shift "pair-datatype.kl" 1))
+(import (shift "list-datatype.kl" 1))
+
+(meta
+  (define identifier?
+    (lambda (stx)
+      (syntax-case stx [(ident x) (true)] [_ (false)]))))
+
+(meta
+  (datatype (Maybe A)
+    (nothing)
+    (just A)))
+
+(meta
+  (defun syntax->list (stx)
+    (syntax-case stx
+      [() (nil)]
+      [(cons x xs) (:: x (syntax->list xs))])))
+
+(meta
+  (define else-case
+     (lambda (stx)
+       (syntax-case stx
+         [(list (e x))
+          (if (identifier? e)
+              (free-identifier-case e
+                [else (pure (just x))]
+                [(else other) (pure (nothing))])
+              (pure (nothing)))]
+         [(ident e)
+          (free-identifier-case e
+            [else (pure (just 'else-x))]
+            [(else other) (pure (nothing))])]
+         [_ (pure (nothing))]))))
+
+(meta
+  (defun list->syntax (stx-lst stx)
+    (case stx-lst
+      [(nil) (empty-list-syntax stx)]
+      [(:: x xs) (cons-list-syntax x (list->syntax xs stx) stx)])))
+
+(meta
+  (datatype (Abstracted)
+    (none)
+    (bind Syntax Syntax (Abstracted)))
+
+  (defun abstract (stxs)
+    (syntax-case stxs
+      [() (none)]
+      [(cons e es)
+       (bind 'internal-x e (abstract es))])))
+   
+
+(define-macros
+  ([pmatch
+    (lambda (stx)
+      (syntax-case stx
+        [(cons _ more)
+         (syntax-case more
+           [(cons scrut pats)
+            (pure `(let (x ,scrut) (pmatch-aux ,scrut x ,pats)))])]))]
+   [pmatch-aux
+    (lambda (stx)
+      (syntax-case stx
+        [(list (_ scrut tgt pats))
+         (syntax-case pats
+           [() (pure `(error ',scrut))]
+           [(cons c cs)
+            (syntax-case c
+              [(list (pat rhs))
+               (>>= (else-case pat)
+                    (lambda (e)
+                      (case e
+                        [(just x) (pure `(let (,x ,tgt) ,rhs))]
+                        [(nothing)
+                         (pure `(let (fk (lambda (_) (pmatch-aux ,scrut ,tgt ,cs)))
+                                  (ppat ,tgt ,pat ,rhs (fk 'hi))))])))])])]))]
+   [ppat
+    (lambda (stx)
+      (syntax-case stx
+        [(list (_ tgt pat ks kf))
```

Comment:
earlier the continuation function was called `fk` and now it's called `kf`. which is it?

Reply:
To be pedantic, the earlier thing is a function, while this is a local name for a piece of code. But I'll call them all "kf".